pkg/common/rotationutil/rotationutil.go

```diff
@@ -22,37 +28,46 @@ func X509Expired(now time.Time, cert *x509.Certificate) bool {
 // JWTSVIDExpiresSoon determines if the given JWT SVID should be rotated
 // based on presented current time, the JWT's expiration.
 // Also returns true if the JWT is already expired.
-func JWTSVIDExpiresSoon(svid *client.JWTSVID, now time.Time) bool {
+func JWTSVIDExpiresSoon(svid *client.JWTSVID, now time.Time, availabilityTarget time.Duration, log logrus.FieldLogger) bool {
 	if JWTSVIDExpired(svid, now) {
 		return true
 	}
 
-	// if the SVID has less than half of its lifetime left, consider it
-	// as expiring soon
-	return shouldRotate(now, svid.IssuedAt, svid.ExpiresAt)
+	// if the SVID has less than half of its lifetime left or reaches the availability target,
+	// consider it as expiring soon
+	return shouldRotate(now, svid.IssuedAt, svid.ExpiresAt, availabilityTarget, log)
 }
 
 // JWTSVIDExpired returns true if the given SVID is expired.
 func JWTSVIDExpired(svid *client.JWTSVID, now time.Time) bool {
 	return !now.Before(svid.ExpiresAt)
 }
 
-func shouldRotate(now, beginTime, expiryTime time.Time) bool {
+func shouldRotate(now, beginTime, expiryTime time.Time, availabilityTarget time.Duration, log logrus.FieldLogger) bool {
 	ttl := expiryTime.Sub(now)
 	// return true quickly if the expiry is already met.
 	if ttl <= 0 {
 		return true
 	}
 
-	halfLife := halfLife(beginTime, expiryTime)
+	if shouldFallbackDefaultRotationStrategy(beginTime, expiryTime, availabilityTarget) {
+		// calculate a jitter delta to spread out rotations
+		halfLife := halfLife(beginTime, expiryTime)
+		delta := jitterDelta(halfLife)
+		min := halfLife - delta
+		jitteredHalfLife := time.Duration(rand.Int63n(int64(delta)*2) + int64(min)) //nolint // gosec: no need for cryptographic randomness here
 
-	// calculate a jitter delta to spread out rotations
-	delta := jitterDelta(halfLife)
-	min := halfLife - delta
+		if ttl <= jitteredHalfLife {
+			if availabilityTarget > 0 {
+				log.Warn("SVID lifetime isn't long enough to guarantee the availability_target, fell back to the default rotation strategy")
+			}
```

Comment:
it looks to me like this log must not be here, since this will happen only when logic from `shouldFallbackDefaultRotationStrategy` return true because the fracePreriod, 
maybe you can move this log there? but that will result in spamming messages...

Reply:
I also think that the log should be done in the `shouldFallbackDefaultRotationStrategy` function. The current implementation already logs this every time that the SVID is being rotated. I'm not sure if I would suggest to log it only once. That could cause to remain unnoticed.

Reply:
Do you think it is OK if the warning log for each entry to be output at every intervals?

> but that will result in spamming messages...
Right, I thought to suppress it.

e.g.
```
WARN[0005] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0005] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0005] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0005] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0009] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0009] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0009] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0010] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0014] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0014] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0014] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0014] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0019] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0019] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0019] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0019] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0024] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0024] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0024] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0024] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0028] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0029] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0029] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0029] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0033] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0034] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0034] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0034] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0038] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0039] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0039] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0039] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0043] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0044] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0044] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0044] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0048] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/spire/agent/join_token/1e6579a0-82c8-4117-bf4c-cc6fcda4780d" subsystem_name=manager
WARN[0049] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0049] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0049] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  spiffe_id="spiffe://example.org/workload" subsystem_name=manager
...
```

Reply:
I dont like that... so you can maybe keep the log where it is, or refactor to simplify it, I think that code can be affected for some of the another commands

Reply:
How about this approch? 
When creating/renewing SVIDs, to check if the availability_target is enough or not.

```
INFO[0000] SVID is not found. Starting node attestation  subsystem_name=attestor trust_domain_id="spiffe://example.org"
INFO[0000] Node attestation was successful               rettestable=false spiffe_id="spiffe://example.org/spire/agent/join_token/f42c9691-21f9-4c9c-a1d0-388e3b60be4b" subsystem_name=attestor trust_domain_id="spiffe://example.org"
DEBU[0000] Entry created                                 entry=50c1d620-74b2-4bf3-887e-b6585a08b84f selectors_added=1 spiffe_id="spiffe://example.org/workload" subsystem_name=cache_manager
DEBU[0000] Entry created                                 entry=f886c54d-775f-4d21-9013-082ef8bdef6d selectors_added=1 spiffe_id="spiffe://example.org/token" subsystem_name=cache_manager
DEBU[0000] Entry created                                 entry=71d139d2-69a3-4ac7-a9b9-8d482a9fcb31 selectors_added=1 spiffe_id="spiffe://example.org/test-workload" subsystem_name=cache_manager
DEBU[0000] Renewing stale entries                        cache_type=workload count=3 limit=500 subsystem_name=manager
INFO[0000] Creating X509-SVID                            entry_id=50c1d620-74b2-4bf3-887e-b6585a08b84f spiffe_id="spiffe://example.org/workload" subsystem_name=manager
INFO[0000] Creating X509-SVID                            entry_id=f886c54d-775f-4d21-9013-082ef8bdef6d spiffe_id="spiffe://example.org/token" subsystem_name=manager
INFO[0000] Creating X509-SVID                            entry_id=71d139d2-69a3-4ac7-a9b9-8d482a9fcb31 spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0000] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  entry_id=f886c54d-775f-4d21-9013-082ef8bdef6d spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0000] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  entry_id=71d139d2-69a3-4ac7-a9b9-8d482a9fcb31 spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0000] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  entry_id=50c1d620-74b2-4bf3-887e-b6585a08b84f spiffe_id="spiffe://example.org/workload" subsystem_name=manager
DEBU[0000] SVID updated                                  entry=f886c54d-775f-4d21-9013-082ef8bdef6d spiffe_id="spiffe://example.org/token" subsystem_name=cache_manager
DEBU[0000] SVID updated                                  entry=71d139d2-69a3-4ac7-a9b9-8d482a9fcb31 spiffe_id="spiffe://example.org/test-workload" subsystem_name=cache_manager
DEBU[0000] SVID updated                                  entry=50c1d620-74b2-4bf3-887e-b6585a08b84f spiffe_id="spiffe://example.org/workload" subsystem_name=cache_manager
DEBU[0000] Bundle added                                  subsystem_name=svid_store_cache trust_domain_id=example.org
INFO[0000] Starting Workload and SDS APIs                address=/tmp/spire-agent/public/api.sock network=unix subsystem_name=endpoints
DEBU[0000] Initializing health checkers                  subsystem_name=health
DEBU[0812] Updating expiring SVIDs in cache              cache_type=workload expiring_svids=1 subsystem_name=manager
DEBU[0812] Renewing stale entries                        cache_type=workload count=1 limit=500 subsystem_name=manager
INFO[0812] Renewing X509-SVID                            entry_id=50c1d620-74b2-4bf3-887e-b6585a08b84f expires_at="2023-11-20T01:13:15Z" spiffe_id="spiffe://example.org/workload" subsystem_name=manager
WARN[0812] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  entry_id=50c1d620-74b2-4bf3-887e-b6585a08b84f spiffe_id="spiffe://example.org/workload" subsystem_name=manager
DEBU[0812] SVID updated                                  entry=50c1d620-74b2-4bf3-887e-b6585a08b84f spiffe_id="spiffe://example.org/workload" subsystem_name=cache_manager
DEBU[0827] Updating expiring SVIDs in cache              cache_type=workload expiring_svids=1 subsystem_name=manager
DEBU[0827] Renewing stale entries                        cache_type=workload count=1 limit=500 subsystem_name=manager
INFO[0827] Renewing X509-SVID                            entry_id=f886c54d-775f-4d21-9013-082ef8bdef6d expires_at="2023-11-20T01:13:15Z" spiffe_id="spiffe://example.org/token" subsystem_name=manager
WARN[0827] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  entry_id=f886c54d-775f-4d21-9013-082ef8bdef6d spiffe_id="spiffe://example.org/token" subsystem_name=manager
DEBU[0827] SVID updated                                  entry=f886c54d-775f-4d21-9013-082ef8bdef6d spiffe_id="spiffe://example.org/token" subsystem_name=cache_manager
DEBU[0832] Updating expiring SVIDs in cache              cache_type=workload expiring_svids=1 subsystem_name=manager
DEBU[0832] Renewing stale entries                        cache_type=workload count=1 limit=500 subsystem_name=manager
INFO[0832] Renewing X509-SVID                            entry_id=71d139d2-69a3-4ac7-a9b9-8d482a9fcb31 expires_at="2023-11-20T01:13:15Z" spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
WARN[0832] SVID lifetime isn't long enough to guarantee the availability_target, fall back to the default rotation strategy  entry_id=71d139d2-69a3-4ac7-a9b9-8d482a9fcb31 spiffe_id="spiffe://example.org/test-workload" subsystem_name=manager
DEBU[0832] SVID updated                                  entry=71d139d2-69a3-4ac7-a9b9-8d482a9fcb31 spiffe_id="spiffe://example.org/test-workload" subsystem_name=cache_manager
DEBU[0850] Rotating agent SVID                           subsystem_name=manager
INFO[0850] Successfully rotated agent SVID               spiffe_id="spiffe://example.org/spire/agent/join_token/f42c9691-21f9-4c9c-a1d0-388e3b60be4b" subsystem_name=manager
```